### PR TITLE
refactor(api): add compatible modules to slot definitions and load by default slot

### PR DESCRIPTION
**Pull request by b-cooper**
## overview

This adds `compatibleModules` to the slots within the deck schema and definitions within shared-data. The addition will allow us to be programmatic about slot/module compatibility as the robot server grows and changes.

The behavior of `protocol.load_module` in PAPIv2 is now slightly augmented.  It will behave the same as it has in the past for TD and MD (e.g. `protocol.load_module('Magnetic Module', 4)`).
Though the Thermocycler Module should now be loaded using `protocol.load_module('Thermocycler Module')` or `protocol.load_module('thermocycler')` with *no specified slot location*. Because the Thermocycler Module has a single compatible slot in the OT2 standard deck definition ('Slot 7'), this will load it into it's position unambiguously.

Similarly, you can load the TC explicitly using `protocol.load_module('thermocycler', 7)`, but it is preferred to use the implicit load for modules with a single compatible slot.

## changelog

- Add listing of compatible modules to deck definitions and schema.
- Allow modules with only one valid slot to be implicitly loaded.

## review requests

- try uploading a TC protocol using the implicit load (`protocol.load_module('thermocycler')`)
- try uploading other modules using explicit slot param (unchanged from before)

### CODE REVIEW

File: api/src/opentrons/protocol_api/contexts.py [hunk 1]
```diff
@@ -250,36 +253,79 @@ def load_labware_by_name(
...
+                return location
+            else:
+                raise AssertionError(
+                    f'module {module_name} cannot be loaded'
+                    ' into slot {location}')
+        else:
+            valid_slots = [slot['id'] for slot in slots if module_name
+                           in slot['compatibleModules']]
+            if len(valid_slots) == 1:
+                pprint(valid_slots)
```
Comment (sfoster1) [inline api/src/opentrons/protocol_api/contexts.py]: Debug prints left in?

File: api/src/opentrons/protocol_api/contexts.py [hunk 2]
```diff
@@ -250,36 +253,79 @@ def load_labware_by_name(
...
+            location: Optional[types.DeckLocation] = None) -> ModuleTypes:
+
         try:
-            geometry = load_module(
-                mod_id, self._deck_layout.position_for(location))
+            resolved_name = self._resolve_module_name(module_name)
+            resolved_location = self._resolve_module_location(
+                    resolved_name, location)
+            location_pos = self._deck_layout.position_for(resolved_location)
+            pprint(location_pos)
```
Comment (sfoster1) [inline api/src/opentrons/protocol_api/contexts.py]: debug prints left in?

File: api/src/opentrons/protocol_api/contexts.py [hunk 3]
```diff
@@ -250,36 +253,79 @@ def load_labware_by_name(
...
+            geometry = load_module(resolved_name, location_pos)
         except KeyError:
```
Comment (sfoster1) [inline api/src/opentrons/protocol_api/contexts.py]: Which call do we expect to raise the keyerror? can we narrow down the scope of the `try` block a little?

File: api/src/opentrons/protocol_api/contexts.py [hunk 4]
```diff
@@ -250,36 +253,79 @@ def load_labware_by_name(
...
+    def _resolve_module_name(self, module_name: str) -> str:
+        alias_map = {
+            'magnetic module': 'magdeck',
+            'temperature module': 'tempdeck',
+            'thermocycler module': 'thermocycler'
+        }
+        lower_name = module_name.lower()
+        return alias_map.get(lower_name, lower_name)
+
+    def _resolve_module_location(
```
Comment (sfoster1) [inline api/src/opentrons/protocol_api/contexts.py]: Let's make this a public function in `geometry` or a public method of `geometry.Deck`, it doesn't seem to use `self` anywhere and the context classes are already pretty large

File: api/src/opentrons/protocol_api/contexts.py [hunk 5]
```diff
@@ -250,36 +253,79 @@ def load_labware_by_name(
...
+            self, module_name: str,
+            location: Optional[types.DeckLocation]) -> types.DeckLocation:
+        # TODO: support deck loadName as a param
+        def_path = 'shared_data/deck/definitions/1/ot2_standard.json'
+        deck_def = json.loads(  # type: ignore
+            pkgutil.get_data('opentrons', def_path))
+        slots: List[Dict] = deck_def['locations']['orderedSlots']
+        if isinstance(location, str) or isinstance(location, int):
```
Comment (sfoster1) [inline api/src/opentrons/protocol_api/contexts.py]: This part from line 274-278 strikes me as something that's going to be used a lot and it's surprisingly complex for how concise it is, maybe we can make it its own function?

File: api/src/opentrons/protocol_api/contexts.py (hunk 3 above)
Comment (sfoster1) [reply api/src/opentrons/protocol_api/contexts.py]: It also looks like the exceptions would be a bit weird here - the only thing in that scope that can raise a `KeyError` specifically is `position_for`, so in most cases you're going to be getting an `AssertionError` or a `ValueError` that bypasses this scope - can we scrub through and make sure that common errors like mispelling a module name, typoing a slot, or specifying an invalid slot raise user friendly errors

File: api/src/opentrons/protocol_api/contexts.py (hunk 1 above)
Comment (b-cooper) [reply api/src/opentrons/protocol_api/contexts.py]: whoops

**Pull request merged**